### ISSUE TITLE
ENH: Add Sheather-Jones bandwidth for KDE within histograms

### ISSUE THREAD

**Issue by qzyu999** [scipy/scipy#19330]
### Is your feature request related to a problem? Please describe.

Currently, I'm not aware whether it's possible to add custom bandwidths for the KDE for histograms.

### Describe the solution you'd like.

I'd like there to be optional bandwidths, namely those listed [https://www.kaggle.com/code/yuqizheng/intro-to-kernel-density-estimation-kde](here). In particular, Sheather-Jones is said to have the best performance empirically. 

### Describe alternatives you've considered.

I believe the only alternative is not Sheather-Jones itself, but from another paper called the Improved Sheather-Jones which I don't know much about.

### Additional context (e.g. screenshots, GIFs)

I learned these from Givens, G. H. G., & Hoeting, J. A. H. (2013). Computational Statistics (2nd ed.). Hoboken, NJ, United States: Wiley. The derivation for Sheather Jones is difficult, but I have the solution in the link provided, along with R-code for calculating it. There are libraries in R which do the same.

**Comment by mdhaber** [comment]
Did you independently write all the R code you referred to, and if so, are you willing to distribute a Python version under SciPy's license? (The license of most R code is some form of the GPL, which is not compatible with that of SciPy.)

I see in https://github.com/statsmodels/statsmodels/issues/7205#issuecomment-1712722883 the suggestion that it be added to stats models. As a more advanced method, do you think it should be in statsmodels instead? Also, is the need not met by the Python libraries that already offer it? (e.g. KDEpy)?

If you are still interested in submitting a PR, I'd suggest reaching out to the mailing list to find a reviewer before getting started. 
https://projects.scipy.org/scipylib/mailing-lists.html
Otherwise, there is a chance that a PR would not be reviewed/merged in reasonable time.

**Comment by qzyu999** [comment]
Yes I did write the code myself in R, and I'd be willing to distribute a Python version with the SciPy license. To be honest, I think that based on the textbook I read since it should be the go-to KDE method I think all libraries should have it.

Quick question, does it being included in SciPy prevent it from being allowed in stats models (under licensing requirements)?

**Comment by mdhaber** [comment]
No, but it probably wouldn't need to be in statsmodels if it were in SciPy because SciPy is a dependancy of statsmodels.

**Comment by e-pet** [comment]
FWIW, here is another quite comprehensive post that claims SJ should generally be the default: https://aakinshin.net/posts/kde-bw/

**Comment by josiahcarlson** [comment]
FWIW, I recently put together an "as simple as I can make it" improved Sheather-Jones implementation starting with just numpy and existing software (translated the zfit reference over in statsmodels. That could / should make it much easier to adapt to scipy:
https://gist.github.com/josiahcarlson/638f9d65b6e3bbb2996ea631ed1ef094

If someone points me to the proper lib in this repo (I'm not a current user of scipy), I can make an effort to port my gist.

**Comment by qzyu999** [comment]
Hi @mdhaber, I've added a PR for this: #24733 in addition to opening a forum post asking for help with a review: https://discuss.scientific-python.org/t/contributor-seeking-reviewer-for-sheather-jones-bandwidth-in-scipy-stats-gaussian-kde/2252

Any guidance on next steps?

**Comment by mdhaber** [comment]
Hmm, no, not much else to suggest. Just to help set expectations, here are a few reasons that responses might be delayed:

- `stats.gaussian_kde` is [not looking like](https://github.com/scipy/scipy/issues/20544) it will be translated to the array API, and that's where a lot of maintainer attention is going these days.
- `stats.gaussian_kde` is not competitive (or designed to be competitive) with other permissively-licensed, lightweight (depending only on SciPy), well-maintained packages like [KDEpy](https://github.com/scipy/scipy/pull/24583#pullrequestreview-3821660783) when it comes to 1D KDE, so there may not be interest in improving its capabilities w.r.t. 1D KDE. (And SciPy doesn't *want* to compete with these libraries. `gaussian_kde` was written early, before these libraries existed, to fulfill a need. But over time, there were still needs that `gaussian_kde` did not fill, and these libraries came about. If the needs are now filled otherwise, we would not want to duplicate the effort.)
- `stats.gaussian_kde` is designed for multidimensional KDE and supports weights, so folks may not be comfortable adding an option that is incompatible with existing options. (It's much better if all options play nicely together.)
- You mention that the code was translated by AI, so it is not immediately obvious that it will be considered compatible with our license, even if you wrote the original code. For instance, we still have not come to a firm conclusion about some code I had AI translate a while ago: https://github.com/scipy/scipy/pull/22298#issuecomment-2623363734. For instance, someone at least would have to look out for whether idioms from license-incompatible code has crept in. (Maybe it's obvious that there's no problem; I'm just saying it's a consideration.)

Unfortunately, I would not personally be interested in reviewing for these reasons (and https://github.com/scipy/scipy/pull/24583#pullrequestreview-3821660783) - that's why I suggested finding a reviewer before getting started. But thanks for your interest, and good luck finding a reviewer!

**Comment by qzyu999** [comment]
Thanks for the follow-up @mdhaber, I understand your points. On a side-note, I did review the AI code and ran tests with it to ensure it matched my previous R code's output.

May you please un-hide my post that was automatically flagged here? https://discuss.scientific-python.org/t/contributor-seeking-reviewer-for-sheather-jones-bandwidth-in-scipy-stats-gaussian-kde/2252

I am hoping I can find a reviewer for this PR, if not I understand.

**Comment by mdhaber** [comment]
I'm not sure why it was flagged, and I don't think I can unhide it. @j-bowhay do you happen to know?

**Comment by qzyu999** [comment]
@mdhaber, @j-bowhay, the post has been restored after a mod took a look, apparently I provided too many links as a new poster and the system flagged me, thanks for the help!
